From ac7cae74e999e667ac7281293a3d07b4aa39b7fe Mon Sep 17 00:00:00 2001
From: ExactExampl <exactxmpl@pixelexperience.org>
Date: Tue, 11 Oct 2022 12:38:00 +0300
Subject: [PATCH 07/11] Conditionally revert "Block Binder thread until
 incoming call process completes"

* Legacy IMS packages handling incoming calls in such a way that
  a blocked binder thread won`t allow to complete call setup,
  thus we have half dead incoming calls with unattached call session
  (caller can hear dialing tone whereas recipient got nothing)

This conditionally reverts commit 75c3dc9ba272b43971f519caba0382f9871c7d9d.

[lostark13: Forward port to Android 14]

Co-Authored-By: Lostark13 <tushar9915@gmail.com>
Change-Id: I55a8f3bbca4a2b9a6bc7511e9fe2d0884a8818e5
---
 .../telephony/imsphone/ImsPhoneCallTracker.java      | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/src/java/com/android/internal/telephony/imsphone/ImsPhoneCallTracker.java b/src/java/com/android/internal/telephony/imsphone/ImsPhoneCallTracker.java
index e397bdc397..ab11fe1fc1 100644
--- a/src/java/com/android/internal/telephony/imsphone/ImsPhoneCallTracker.java
+++ b/src/java/com/android/internal/telephony/imsphone/ImsPhoneCallTracker.java
@@ -390,7 +390,17 @@ public class ImsPhoneCallTracker extends CallTracker implements ImsPullCall {
         @Nullable
         public IImsCallSessionListener onIncomingCall(
                 @NonNull IImsCallSession c, @Nullable String callId, @Nullable Bundle extras) {
-            return executeAndWaitForReturn(()-> processIncomingCall(c, callId, extras));
+            final boolean shouldBlockBinderThreadOnIncomingCalls = SystemProperties.getBoolean(
+                    "ro.telephony.block_binder_thread_on_incoming_calls", true);
+            if (shouldBlockBinderThreadOnIncomingCalls) {
+                return executeAndWaitForReturn(()-> processIncomingCall(c, callId, extras));
+            } else {
+                // for legacy IMS we want to avoid blocking the binder thread, otherwise
+                // we end up with half dead incoming calls with unattached call session
+                TelephonyUtils.runWithCleanCallingIdentity(()-> processIncomingCall(
+                        c, callId, extras), mExecutor);
+                return null;
+            }
         }
 
         @Override
-- 
2.39.2

